File: v1/Python/beginner/05_ShortestPath/boj_13549.py
import sys
import heapq

input = sys.stdin.readline

def solve(grp: list[list[tuple[int]]], start: int, end: int) -> int:
    q = []
    distance = [int(1e9)] * (len(grp) + 1)

    heapq.heappush(q, (0, start))
    distance[start] = 0

    while q:
        dist, now = heapq.heappop(q)

        if distance[now] < dist:
            continue

        for tup in grp[now]:
            for t in tup:
                if t >= max(start, end) * 2:
                    continue

                weight = 0 if now * 2 == t else 1
                cost = distance[now] + weight

                if distance[t] > cost:
                    distance[t] = cost
                    heapq.heappush(q, (cost, t))

    return distance[end]

if __name__ == "__main__":
    n, k = map(int, input().split())

    # 모든 i번 노드에는 i - 1, i + 1, 2i 노드가 연결되어 있다. 
    # i - 1과 i + 1은 가중치가 1이고, 2i은 가중치가 0이다.

    m = max(n, k)

    grp = [[] for _ in range((2 * m) + 1)]
    grp[0] = [(0, 1)]

    for i in range(1, len(grp)):
        if i + 1 == 2 * i:
            grp[i].append((i - 1, i + 1))
        else:
            grp[i].append((i - 1, i + 1, 2 * i))

    print(solve(grp, n, k))
     

File: v1/Java/boj/Beginner/E_ShortestPath/boj_13549.java
package boj.Beginner.E_ShortestPath;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;

public class boj_13549 {
    static StringTokenizer st;
    static BufferedReader br =
            new BufferedReader(new InputStreamReader(System.in));

    static int n, k, m;
    static List<Integer>[] grp;

    static class Node implements Comparable<Node> {
        int end, dist;

        public Node(int end, int dist) {
            this.end = end;
            this.dist = dist;
        }

        @Override
        public int compareTo(Node o) {
            return this.dist - o.dist;
        }
    }

    private static int solve() {
        PriorityQueue<Node> pq = new PriorityQueue<>();
        int[] distance = new int[grp.length];
        Arrays.fill(distance, Integer.MAX_VALUE);

        distance[n] = 0;
        pq.add(new Node(n, 0));

        while (!pq.isEmpty()) {
            Node tmp = pq.poll();
            int now = tmp.end;
            int dist = tmp.dist;

            if (distance[now] < dist)
                continue;

            for (int k : grp[now]) {
                if (k >= 2 * m)
                    continue;

                int weight = k == now * 2 ? 0 : 1;
                int cost = distance[now] + weight;

                if (distance[k] > cost) {
                    distance[k] = cost;
                    pq.add(new Node(k, cost));
                }
            }
        }

        return distance[k];
    }

    public static void main(String[] args) throws IOException {
        st = new StringTokenizer(br.readLine());
        n = Integer.parseInt(st.nextToken());
        k = Integer.parseInt(st.nextToken());
        m = Integer.max(n, k);

        grp = new ArrayList[2 * m];
        Arrays.setAll(grp, elem -> new ArrayList<>());

        grp[0].add(1);
        grp[1].add(2);
        grp[1].add(0);

        for (int i = 2; i < grp.length; i++) {
            grp[i].add(i - 1);
            grp[i].add(i + 1);
            grp[i].add(2 * i);
        }

        System.out.print(solve());
        br.close();
    }
}
